test/cpp/microbenchmarks/callback_unary_ping_pong.h

```diff
@@ -0,0 +1,84 @@
+/*
+ *
+ * Copyright 2016 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/* Benchmark gRPC end2end in various configurations */
+
+#ifndef TEST_CPP_MICROBENCHMARKS_CALLBACK_UNARY_PING_PONG_H
+#define TEST_CPP_MICROBENCHMARKS_CALLBACK_UNARY_PING_PONG_H
+
+#include <benchmark/benchmark.h>
+#include <sstream>
+#include "src/core/lib/profiling/timers.h"
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/cpp/microbenchmarks/fullstack_context_mutators.h"
+#include "test/cpp/microbenchmarks/fullstack_fixtures.h"
+#include "test/cpp/microbenchmarks/callback_test_service.h"
+
+namespace grpc {
+namespace testing {
+
+/*******************************************************************************
+ * BENCHMARKING KERNELS
+ */
+
+template <class Fixture, class ClientContextMutator, class ServerContextMutator>
+static void BM_CallbackUnaryPingPong(benchmark::State& state) {
+  CallbackStreamingTestService service;
+  std::unique_ptr<Fixture> fixture(new Fixture(&service));
+  std::unique_ptr<EchoTestService::Stub> stub_(
+      EchoTestService::NewStub(fixture->channel()));
+  EchoRequest request;
+  EchoResponse response;
+
+  if (state.range(0) > 0) {
+    request.set_message(std::string(state.range(0), 'a'));
+  } else {
+    request.set_message("");
+  }
+  if (state.range(1) > 0) {
+    response.set_message(std::string(state.range(1), 'a'));
+  } else {
+    response.set_message("");
+  }
+
+  while (state.KeepRunning()) {
```

Comment:
As I mentioned when we spoke, I would suggest changing this to if (state.KeepRunning()) and then move the contents here to a new function.

Reply:
When I try to move the content to a new function and call it recursively, I need to check if(state.KeepRunning()) inside the function to stop the recursion. Code looks like this:
`static void BM_CallbackUnaryPingPong(benchmark::State& state) {
......
 if (state.KeepRunning()) {
    GPR_TIMER_SCOPE("BenchmarkCycle", 0);
    SendCallbackUnaryPingPong(state, request, response, stub_);
  }
.....
}

void SendCallbackUnaryPingPong(::benchmark::State state, EchoRequest request,
                               EchoResponse response,
                               std::unique_ptr<EchoTestService::Stub> stub_) {
  if (state.KeepRunning()) {
     ClientContext cli_ctx;
     stub_->experimental_async()->Echo(&cli_ctx, &request, &response,
                                       [state, &request, &response, stub_](Status s) {
       GPR_ASSERT(s.ok());
       SendCallbackUnaryPingPong(state, request, response, stub_);
     });
   }
}`


There are two major errors I got, first is that the recursive function is deleted when calling stub_->experimental_async()->Echo(), because of std::move(f) is implemented inside Echo(). So it gives me error like this:
`/usr/include/c++/7/bits/std_function.h:268:39: error: use of deleted function ‘grpc::testing::SendCallbackUnaryPingPong(benchmark::State, grpc::testing::EchoRequest, grpc::testing::EchoResponse, std::unique_ptr<grpc::testing::EchoTestService::Stub>)::<lambda(grpc::Status)>::<lambda>(grpc::testing::SendCallbackUnaryPingPong(benchmark::State, grpc::testing::EchoRequest, grpc::testing::EchoResponse, std::unique_ptr<grpc::testing::EchoTestService::Stub>)::<lambda(grpc::Status)>&&)’
  { __functor._M_access<_Functor*>() = new _Functor(std::move(__f)); }`

Second is that benchmark::State is not allowed to be passed in as an argument. I tried to add benchmark::State to SendCallbackUnaryPingPong as an argument but it keeps giving me this error :
`third_party/benchmark/include/benchmark/benchmark.h:633:38: note: declared private here
   BENCHMARK_DISALLOW_COPY_AND_ASSIGN(State);
                                      ^
third_party/benchmark/include/benchmark/benchmark.h:200:3: note: in definition of macro ‘BENCHMARK_DISALLOW_COPY_AND_ASSIGN’
   TypeName(const TypeName&) = delete;    `

Reply:
I don't have full context on your previous conversation here, but why does this have to be recursive? Couldn't you do something like:

static void BM_CallbackUnaryPingPong(benchmark::State& state) {
......
if (state.KeepRunning()) {
  GPR_TIMER_SCOPE("BenchmarkCycle", 0);
  SendCallbackUnaryPingPong(request, response, stub_);
}
.....
}

void SendCallbackUnaryPingPong(EchoRequest* request, EchoResponse* response,
  std::unique_ptr<EchoTestService::Stub> stub) {
  ClientContext cli_ctx;
  stub_->experimental_async()->Echo(&cli_ctx, request, response,
    [state, request, response, stub](Status s) {
      GPR_ASSERT(s.ok());
      std::lock_guard<std::mutex> l(mu);
      done = true;
      cv.notify_one();
    });

   std::unique_lock<std::mutex> l(mu);
   while (!done) {
      cv.wait(l);
   }
}

Reply:
In terms of your error with passing in the benchmark state as a parameter, it looks like you are trying to copy construct the Benchmark state. Did you try:
'const ::benchmark::State& state'?

Reply:
yeah, I tried const ::benchmark::State& state before and it had the same error: 
./test/cpp/microbenchmarks/callback_unary_ping_pong.h:77:41: error: ‘benchmark::State::State(const benchmark::State&)’ is private within this context
In file included from ./test/cpp/microbenchmarks/callback_unary_ping_pong.h:24:0,
                 from test/cpp/microbenchmarks/bm_callback_unary_ping_pong.cc:19:
third_party/benchmark/include/benchmark/benchmark.h:633:38: note: declared private here
   BENCHMARK_DISALLOW_COPY_AND_ASSIGN(State);
                                      ^
third_party/benchmark/include/benchmark/benchmark.h:200:3: note: in definition of macro ‘BENCHMARK_DISALLOW_COPY_AND_ASSIGN’
   TypeName(const TypeName&) = delete;